src/plonk/mod.rs

```diff
@@ -248,24 +248,39 @@ impl<F: PrimeField> PlonkStructure<F> {
             W1s: &W.W,
             W2s: &vec![],
         };
-        let nrow = 1 << self.k;
-        let res = (0..nrow)
+
+        let total_row = 1 << self.k;
+        (0..total_row)
             .into_par_iter()
-            .map(|row| data.eval(&self.poly, row))
-            .collect::<Result<Vec<F>, _>>()
-            .map(|v| v.into_iter().filter(|v| F::ZERO.ne(v)).count())?;
-
-        let is_h_equal_g = self.is_sat_log_derivative(&W.W);
-
-        match (res == 0, check_commitments == 0, is_h_equal_g) {
-            (true, true, true) => Ok(()),
-            (false, _, _) => Err(Error::EvaluationMismatch {
-                mismatch_count: res,
-                total_row: nrow,
-            }),
-            (_, _, false) => Err(Error::LogDerivativeNotSat),
-            _ => Err(Error::CommitmentMismatch),
+            .map(|row| {
+                data.eval(&self.poly, row)
+                    .map(|row_result| if row_result.eq(&F::ZERO) { 0 } else { 1 })
+            })
+            .try_reduce(
+                || 0,
+                |mismatch_count, is_missed| Ok(mismatch_count + is_missed),
+            )
+            .map(|mismatch_count| {
+                Some(Error::EvaluationMismatch {
+                    mismatch_count: NonZeroUsize::new(mismatch_count)?,
+                    total_row,
+                })
+            })?
+            .err_or(())?;
+
+        if !self.is_sat_log_derivative(&W.W) {
+            return Err(Error::LogDerivativeNotSat);
         }
+
+        U.W_commitments
```

Comment:
We don't need move this to bottom. By keep it in the original place will make the PR diff less.

Reply:
The change of order here is not cosmetic, but is used to keep the variables as close to their place of use as possible so that the logic of the function can be easily followed. 

This rule makes it easier to read functions, the "thought" does not spread over the whole function, but is local. 

Given the size of the PR I think the overhead of looking at it is much less than if I have to do two PRs, where in the second one I just move the expressions to where they are now. 

Let me know if you insist, I'll split it into two PRs

Reply:
Moreover, if earlier we still counted all the checks but ignored all the errors except the first one (and the other errors didn't even get into the logs), now we go sequentially and no meaningless calculations are performed.

Reply:
looks good to me